srml/rolling-window/src/lib.rs

```diff
@@ -0,0 +1,280 @@
+	}
+}
+
+/// Trait for reporting misbehavior's
+pub trait MisbehaviorReporter<Kind, Hash> {
+	/// Report misbehavior for a kind
+	fn report_misbehavior(kind: Kind, footprint: Hash, current_session: SessionIndex) -> Result<(), ()>;
+}
+
+impl<T: Trait> MisbehaviorReporter<T::Kind, T::Hash> for Module<T> {
+	fn report_misbehavior(
+		kind: T::Kind,
+		footprint: T::Hash,
+		current_session: SessionIndex,
+	) -> Result<(), ()> {
+		if <BondingUniqueness<T>>::exists(footprint) {
+			Err(())
+		} else {
+			<BondingUniqueness<T>>::insert(footprint, current_session);
+			<MisbehaviorReports<T>>::mutate(kind, |entry| entry.push(current_session));
+			Ok(())
+		}
+	}
+}
+
+impl<T: Trait> srml_session::OnSessionEnding<T::AccountId> for Module<T> {
+	fn on_session_ending(ending: SessionIndex, _applied_at: SessionIndex) -> Option<Vec<T::AccountId>> {
+		for (kind, _) in <MisbehaviorReports<T>>::enumerate() {
+			let window_length = kind.window_length();
+			<MisbehaviorReports<T>>::mutate(kind, |reports| {
+				// it is guaranteed that `reported_session` happened in the same session or before `ending`
+				reports.retain(|reported_session| {
+					let diff = ending.wrapping_sub(*reported_session);
+					diff < *window_length
+				});
+			});
+		}
+
+		// don't provide a new validator set
+		None
+	}
+}
+
+/// Macro for implement static `base_severity` which may be used for misconducts implementations
+#[macro_export]
+macro_rules! impl_base_severity {
+	// type with type parameters
+	($ty:ident < $( $N:ident $(: $b0:ident $(+$b:ident)* )? ),* >, $t: ty : $seve: expr) => {
+		impl< $( $N $(: $b0 $(+$b)* )? ),* > $ty< $( $N ),* > {
+			fn base_severity() -> $t {
+				$seve
+			}
+		}
+	};
+	// type without type parameters
+	($ty:ident, $t: ty : $seve: expr) => {
+		impl $ty {
+			fn base_severity() -> $t {
+				$seve
+			}
+		}
+	};
+}
+
+/// Macro for implement static `misconduct kind` which may be used for misconducts implementations
+/// Note, that the kind need to implement the `WindowLength` trait to work
+#[macro_export]
+macro_rules! impl_kind {
+	// type with type parameters
+	($ty:ident < $( $N:ident $(: $b0:ident $(+$b:ident)* )? ),* >, $t: ty : $kind: expr) => {
+
+		impl< $( $N $(: $b0 $(+$b)* )? ),* > $ty< $( $N ),* > {
+			fn kind() -> $t {
+				$kind
+			}
+		}
+	};
+	// type without type parameters
+	($ty:ident, $t: ty : $kind: expr) => {
+		impl $ty {
+			fn kind() -> $t {
+				$kind
+			}
+		}
+	};
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use runtime_io::with_externalities;
+	use crate::mock::*;
+	use substrate_primitives::H256;
+	use srml_session::OnSessionEnding;
+
+	type RollingWindow = Module<Test>;
+
+	#[test]
+	fn it_works() {
+		with_externalities(&mut new_test_ext(), || {
+			let zero = H256::zero();
+			let one: H256 = [1_u8; 32].into();
+			let two: H256 = [2_u8; 32].into();
+
+			let mut current_session = 0;
+
+			assert!(RollingWindow::report_misbehavior(Kind::Two, zero, current_session).is_ok());
```

Comment:
```suggestion
			assert_ok!(RollingWindow::report_misbehavior(Kind::Two, zero, current_session));
```

this is better since it'll nicely print the error text in tests if it fails.

Reply:
should be replicated in rest of tests